### ISSUE TITLE
GUI

### ISSUE THREAD

**Issue by raboof** [raboof/nethogs#13]
It would be nice if the data in nethogs could be navigated through a GUI. 

This issue exists to track ideas/progress on different approaches to do this.

**Comment by raboof** [comment]
One approach is using Electron, discussed in #4

**Comment by mb-gh** [comment]
Hi Guys,

I am considering implementing a GUI for nethogs.
The idea is to make nethogs supports a daemon mode and then create a separate GUI that will register to receive updates from the daemon when needed.
The GUI would be developed in GTK.
For the communication part, I would think D-BUS is a good choice but I never used it before so If you have another recommendation, please tell me.
Cheers.
Mohamed

**Comment by raboof** [comment]
That'd be great!

Yes, having a 'daemon mode' and separating the GUI as a separate application seems like a reasonable way to go. 

I'm not too sure about D-BUS, as it adds the requirement of running a dbus-daemon, which appears like a fairly complicated thing. I wonder if we could get some inspiration from the realtime/IoT/time-series-database world?

**Comment by mb-gh** [comment]
D-BUS Seems to be a bit complicated indeed.
You want may have a look at [this](http://stackoverflow.com/questions/2281204/which-linux-ipc-technique-to-use) question on stackoverflow.
They suggest Unix domain sockets

**Comment by mb-gh** [comment]
Here is an example of Unix-Domain-Sock client/server https://github.com/troydhanson/network/tree/master/unixdomain

**Comment by raboof** [comment]
Unix domain sockets would work for me, though 'normal' network sockets would be fine too.

Question remains what kind of protocol we should use on top of that socket, it would be nice to use some kind of 'standard' without pulling in too much complexity :).

**Comment by mb-gh** [comment]
The advantages of Unix domain sockets over IP sockets is that they are very fast and do not need you to define a port number.

About the data format, I  thought about sending simple C structures initially with some ID as first field to differentiate them. It should not be complex and should be fairy easy to write and parse because we just cast to data to the right struct.

**Comment by raboof** [comment]
I don't expect we'd send such volumes of data that the difference between unix domain sockets and IP sockets becomes noticable ;).

Indeed for unix domain sockets you don't need a port number, but you do need a filesystem inode, so that seems kind of equivalent.

Sending C structures over does limit the kind of clients that can (easily) consume the API.

On the other hand: I don't want to block you, if you can get a nice GUI going with unix domain sockets and sending C structures over I'd say go for it! :)

**Comment by mb-gh** [comment]
I will see what I can do :)

**Comment by mb-gh** [comment]
I just discovered that you can enable libcap capabilities for non root process by running the following as root **only once**

> setcap cap_net_raw,cap_net_admin=eip <process path>

I think this is super interesting because we can also consider making a libnethog from the currently code and link to the GUI app without having any inter-process communication.

I excluded this idea initially because I did **not** like having to type root password every-time the user starts the GUI (and did not know about the setcap trick above).

Considering this, the library way may be easier. I will do an attempt and see if it works.

**Comment by raboof** [comment]
I'd certainly be open to linking in nethogs as a library instead of doing IPC.

I wouldn't be comfortable recommending `setcap` to users until we've done a good assessment of the security implications, though, but that's a separate issue ;)

**Comment by raboof** [comment]
you could even consider building 2 binaries from the same sources, without making nethogs a 'real' shared library

**Comment by mb-gh** [comment]
Yes. That could be done too.
But then that will become a separate project that just steals the libcap part from nethogs ;)

**Comment by mb-gh** [comment]
If we want to share code, we have to share a library. If not we have to copy the 'core' part of nethogs in a new project. what do you think?

**Comment by raboof** [comment]
I agree we don't want to have to copy the nethogs 'core' into a new project.

Including the GUI inside the nethogs project would not be necessarily unacceptable to me, though. The obvious downside is this gets messy when there's multiple 'competing' GUI's in the future. I'm leaving that up to you for now :)

**Comment by mb-gh** [comment]
That's the thing... we don't want to force KDE people to use GTK.. they won't be happy. :)

I will attempt to add new build mode to the make file for building a library and then use it in my app.

In the  future, there could be another project that moves the ncurses display part into another separate project doing just console display and keeping the core features in the lib.

**Comment by raboof** [comment]
:+1: !

**Comment by mb-gh** [comment]
The library part is in the final stage.
[(https://github.com/raboof/nethogs/compare/master...mb-gh:build-as-lib)]

I added some new files for the lib and did not change any existing logic.
But I had to move few code blocks between files.

The result gives a clear separation between files:
- Library related files:` libnethogs.cpp libnethogs.h MakeLib.mk`
- NCurses display and main: `main.cpp cui.cpp cui.h`
- Common logic: other files

For even better organisation, I was thinking about moving the lib related files into a `lib` folder and the app related files into an `app` folder, but I was afraid it makes merge difficult for any pending changes.

What do you think? can/should I move those files around?

**Comment by mb-gh** [comment]
So are you interested by my pull request?
If yes, should I created some ticket or how do you proceed with merge?
If you think it's not fit for merge, tell me so we consider other approches.
Thanks

**Comment by raboof** [comment]
Yes, I'm very much interested in your PR! Also very happy to see how the UI is taking shape :)

I do want to be careful before merging, so there will probably be some review rounds before we can merge this to master, but your work is certainly much appreciated. It might take some time for me to find time and dive into this :).

**Comment by mb-gh** [comment]
That's cool.
The GUI will be simplistic but sufficient.
I may not have the have to give something like a preference page etc...
But it should include all the information you can see in the console UI.

**Comment by mb-gh** [comment]
I also need to consider some annoying stuff like adaptability to translation to other languages and the autools files mess :)

**Comment by mb-gh** [comment]
The initial version of the GUI as ready!
https://github.com/mb-gh/gnethogs/wiki

**Comment by raboof** [comment]
Added a link to gnethogs to the README, I suppose we can close this issue
